dream.py

```diff
@@ -144,40 +208,193 @@ def generate_data(self, n_events):
+        the events are drawn from many poisson distributions with an
+        expectation value equal to n_events. For typical n_event numbers,
+        the difference is much less than 1%.
+        
+        Parameters
+        ----------
+            n_events: int
+                Number of events to generate
+        
+            wavelength_width: float [Å]
+                Width of wavelength frame of the instrument, 3.6 Å should fill the frame with perfect source
+            
+            wavelength_center: float [Å]
+                Center of wavelength frame of the instrument, typicall around 2.5-3.0 Å
+            
+            effective_pulse_length: float [s]
+                Effective pulse length (after choppers have modified the pulse), between 10 us and 2.86 ms
+        
+            progress_bar: Boolean
+                Set to True for a progress bar of data generation to be written to terminal
+            
+            verbose: Boolean
+                Set to True for a short analysis of generated data to be written to terminal
         """
-        self.hist = sc.histogram(self.dspacing, dspacing.coords[Dim.DSpacing])
-        return self.hist
+    
+        if wavelength_center - wavelength_width/2.0 < 0:
+            raise ValueError("Provided wavelength band with negative wavelengths.")
+    
+        # get pixel theta:
+        positions = np.array(dream.d.coords[Dim.Position].values)
+        n_positions = (len(positions))
+        z_dir = np.zeros((n_positions, 3))
+        z_dir[:,2] = 1.0
+        position_lengths = np.sqrt(np.sum(positions**2, axis=1))
+        position_zdir_dots = np.sum(positions*z_dir, axis=1)
+        tthetas = 180/np.pi*np.array(np.arccos(position_zdir_dots/position_lengths))
+        thetas = 0.5*tthetas
+        
+        # Find allowed bragg peaks for each pixel and find total strength in order to normalize
+        allowed = np.full((n_positions, len(self.sample_bragg_peaks)), False)
+        pixel_strengths = np.zeros(n_positions)
+        count = 0
+        for bragg_peak in self.sample_bragg_peaks:
+            wavelength = 2.0*bragg_peak.d_spacing*np.sin(thetas)
+            allowed[:,count] = np.logical_and(wavelength < wavelength_center + wavelength_width/2.0,
+                                              wavelength > wavelength_center - wavelength_width/2.0)
+            pixel_strengths[allowed[:,count]] += bragg_peak.strength
+            count += 1
+            
+        # Now we have pixel_strength as indicator for relative intensity in each peak (assuming relatively narrow peaks)
+        total_strength = np.sum(pixel_strengths)
+        events_per_strength = n_events / total_strength
 
-    #def data_reduction(self):
-        # "DiffractionFocussing" == sum? (not available yet)
-        # focussed = sc.sum(hist, Dim.Position)
+        # Calculate distance from source to each pixel when scattered in sample
+        source_to_pixel = self.source_sample_dist + position_lengths
```

Comment:
We have exactly this code (computing scattering angles, L2, ...) in the implementation of the unit conversions in C++. We should discuss the best way of making this available in Python. Either as part of a task in your next sprint, or independently.